js/actions/load-rubric.js

```diff
@@ -48,16 +71,29 @@ export function loadRubric (data) {
 export function requestRubric () {
   return (dispatch, getState) => {
     const state = getState()
-    const urls = state.getIn(['report', 'activities']).map(a => a.get('rubricUrl'))
-    urls.filter(u => u).forEach((u) => {
+    const activities = state.getIn(['report', 'activities'])
+    const withRubric = activities.filter(act => exists(act, 'rubric') && exists(act, 'rubricUrl'))
+    const withoutRubric = activities.filter(act => !exists(act, 'rubric'))
+    const needingFetch = withoutRubric.filter(act => exists(act, 'rubricUrl'))
+
+    withRubric.forEach(a => {
       dispatch({
-        type: REQUEST_RUBRIC,
-        callAPI: {
-          type: 'loadRubric',
-          data: u,
-          successAction: addRubric
-        }
+        type: LOAD_RUBRIC,
+        url: a.get('rubricUrl'),
+        rubric: a.get('rubric')
       })
     })
+
+    needingFetch.map(a => a.get('rubricUrl'))
+      .forEach((u) => {
+        dispatch({
+          type: REQUEST_RUBRIC,
+          callAPI: {
+            type: 'loadRubric',
+            data: u,
+            successAction: addRubric
+          }
+        })
+      })
```

Comment:
Those two API calls, `withRubric` and `needingFetch` are a bit unclear to me. I guess if I analyzed everything better, I could understand it, but maybe some comments could be useful. It's kind of confusing, as one action is LOAD_RUBRIC, other is REQUEST_RUBRIC, which is in fact API call "loadRubric", and so on. 😉

Reply:
Yes – this is a very good point.  Its not obvious what is happening here. It needs some explanation.

Thanks!